packages/core/src/machine.schema.json

```diff
@@ -0,0 +1,316 @@
+                "$ref": "#/definitions/actionObject"
+              }
+            },
+            "invoke": {
+              "$ref": "#/definitions/invokeArray"
+            },
+            "on": {
+              "$ref": "#/definitions/transitionsObject"
+            }
+          },
+          "required": ["on"]
+        }
+      ]
+    },
+    "historyStateNode": {
+      "allOf": [
+        { "$ref": "#/definitions/baseStateNode" },
+        {
+          "type": "object",
+          "properties": {
+            "type": {
+              "type": "string",
+              "pattern": "history"
+            },
+            "history": {
+              "type": "string",
+              "enum": ["shallow", "deep"]
+            }
+          },
+          "required": ["history"]
+        }
+      ]
+    },
+    "finalStateNode": {
+      "allOf": [
+        { "$ref": "#/definitions/baseStateNode" },
+        {
+          "type": "object",
+          "properties": {
+            "type": {
+              "type": "string",
+              "pattern": "final"
+            },
+            "data": {
+              "type": "object"
+            }
+          }
+        }
+      ]
+    },
+    "statesObject": {
+      "type": "object",
+      "patternProperties": {
+        "^.*$": {
+          "oneOf": [
+            { "$ref": "#/definitions/atomicStateNode" },
+            { "$ref": "#/definitions/compoundStateNode" },
+            { "$ref": "#/definitions/parallelStateNode" },
+            { "$ref": "#/definitions/historyStateNode" },
+            { "$ref": "#/definitions/finalStateNode" }
+          ]
+        }
+      }
+    },
+    "transitionsObject": {
+      "type": "object",
+      "patternProperties": {
+        "^.*$": {
+          "type": "array",
+          "items": {
+            "type": "object",
+            "properties": {
+              "actions": {
+                "type": "array",
+                "items": {
+                  "$ref": "#/definitions/actionObject"
+                }
+              },
+              "cond": {
+                "type": "object"
+              },
+              "eventType": {
+                "type": "string"
+              },
+              "source": {
+                "type": "string"
+              },
+              "target": {
+                "type": "array",
+                "items": {
+                  "type": "string"
+                }
+              }
+            },
+            "required": ["actions", "eventType", "source", "target"]
+          }
+        }
+      }
+    },
+    "invokeObject": {
+      "type": "object",
+      "properties": {
+        "type": {
+          "type": "string"
+        },
+        "id": {
+          "type": "string"
+        },
+        "src": {
+          "type": "string"
+        },
+        "autoForward": {
+          "type": "boolean",
+          "default": false
+        }
+      },
+      "required": ["type", "id", "src"],
+      "additionalProperties": false
```

Comment:
I'm trying to test this JSON schema against my implementation. 
I got a validation issue because my `invoke`  contains `onDone` and `onError` as additional properties. 

Error:

```JS
{ keyword: 'additionalProperties',
        dataPath: '.states[\'f941522f-a625-4953-9472-7d3c4004d1d4\'].invoke[0]',
        schemaPath: '#/definitions/invokeObject/additionalProperties',
        params: { additionalProperty: 'onError' },
        message: 'should NOT have additional properties' },
```

Object:
```js
states: {
      [myId]: {
        id: 'doSomething',
        key: myId,
        type: 'atomic',
        invoke: [
          {
            src: 'doSomething',
            onDone: `${myId}_EXECUTED`, 
            onError: `${myId}_FAILED`,
          },
        ],
      }
```
https://xstate.js.org/docs/guides/communication.html#the-invoke-property 
Should you add `onDone` and `onError` in the schema as well or I'm missing something?

Reply:
I'm not sure - but maybe you test the wrong thing? onError & onDone are part of the config but they are converted to transitions internally and thus probably available on the definition (and not the config)

Reply:
I see it makes sense. Indeed, I'm creating a config so I can't test against this JSON schema. I will wait for this feature to get merge then